File: solution/ts/tests/ArticleBuilder.ts
import {faker} from "@faker-js/faker";
import * as E from "fp-ts/Either";
import {Article} from "../src/article";

export const Author = "Pablo Escobar";
export const CommentText = "Amazing article !!!";
export const CreationDate = new Date(2024, 10, 5, 13, 0, 1);

export class ArticleBuilder {
    private comments: Record<string, string> = {};

    static anArticle = (): ArticleBuilder => new ArticleBuilder();

    private static timeProvider: () => Date = () => CreationDate;

    commented(): ArticleBuilder {
        this.comments[CommentText] = Author;
        return this;
    }

    build(): Article {
        let article = Article.create(faker.lorem.sentence(), faker.lorem.paragraph(), ArticleBuilder.timeProvider);
        Object.entries(this.comments)
            .forEach(([text, author]) => {
                let result = article.addComment(text, author);
                if (E.isRight(result)) article = result.right;
            });
        return article;
    }
}

File: solution/python/tests/article_builder.py
from datetime import datetime

from faker import Faker

from blog.article import Article, Comment

AUTHOR = "Pablo Escobar"
COMMENT_TEXT = "Amazing article !!!"
CREATION_DATE = datetime(2024, 11, 5, 13, 0, 1)

fake = Faker()


def time_provider() -> datetime:
    return CREATION_DATE


class ArticleBuilder:
    def __init__(self):
        self._name = fake.name()
        self._content = fake.text()
        self._comments = []
        self._time_provider = time_provider

    @staticmethod
    def an_article() -> 'ArticleBuilder':
        return ArticleBuilder()

    def commented(self) -> 'ArticleBuilder':
        self._comments.append(Comment(COMMENT_TEXT, AUTHOR, self._time_provider()))
        return self

    def build(self) -> Article:
        article = Article.create(self._name, self._content, self._time_provider)
        for comment in self._comments:
            result = article.add_comment(comment.text, comment.author)
            if isinstance(result, Article):
                article = result
        return article
